src/core/emotiondisasm.cpp

```diff
@@ -5,56 +5,1058 @@
+            return disasm_lwl(instruction);
+        case 0x23:
+            return disasm_lw(instruction);
+        case 0x24:
+            return disasm_lbu(instruction);
+        case 0x25:
+            return disasm_lhu(instruction);
+        case 0x26:
+            return disasm_lwr(instruction);
+        case 0x27:
+            return disasm_lwu(instruction);
+        case 0x28:
+            return disasm_sb(instruction);
+        case 0x29:
+            return disasm_sh(instruction);
+        case 0x2A:
+            return disasm_swl(instruction);
+        case 0x2B:
+            return disasm_sw(instruction);
+        case 0x2C:
+            return disasm_sdl(instruction);
+        case 0x2D:
+            return disasm_sdr(instruction);
+        case 0x2E:
+            return disasm_swr(instruction);
+        case 0x2F:
+            return "cache";
+        case 0x31:
+            return disasm_lwc1(instruction);
+        case 0x36:
+            return "TODO: lqc2";
+            break;
+        case 0x37:
+            return disasm_ld(instruction);
+        case 0x39:
+            return disasm_swc1(instruction);
+        case 0x3E:
+            return "TODO: sqc2";
+        case 0x3F:
+            return disasm_sd(instruction);
+        default:
+            return unknown_op("", instruction >> 26, 2);
     }
-    return "[UNKNOWN]";
 }
 
-/*
- * Branch disassembly uses "instr_addr", the address of the instruction in memory, to calculate the
- * new address that can be branched to.
- */
+
+string EmotionDisasm::disasm_j(uint32_t instruction, uint32_t instr_addr)
+{
+    return disasm_jump("j", instruction, instr_addr);
+}
+
+string EmotionDisasm::disasm_jal(uint32_t instruction, uint32_t instr_addr)
+{
+    return disasm_jump("jal", instruction, instr_addr);
+}
+
+
+string EmotionDisasm::disasm_jump(const string opcode, uint32_t instruction, uint32_t instr_addr)
+{
+    stringstream output;
+    uint32_t addr = (instruction & 0x3FFFFFF) << 2;
+    addr += (instr_addr + 4) & 0xF0000000;
+    output << "$" << setfill('0') << setw(8) << hex << addr;
+
+    return opcode + " " + output.str();
+}
+
 string EmotionDisasm::disasm_bne(uint32_t instruction, uint32_t instr_addr)
 {
-    //Use stringstream only when putting integers into the disassembly.
+    return disasm_bequality("bne", instruction, instr_addr);
+}
+
+
+string EmotionDisasm::disasm_bequality(string opcode, uint32_t instruction, uint32_t instr_addr)
+{
     stringstream output;
-    string opcode = "bne";
-    int offset = (int16_t)(instruction & 0xFFFF);
-    offset <<= 2;
-    uint64_t reg1 = (instruction >> 21) & 0x1F;
-    uint64_t reg2 = (instruction >> 16) & 0x1F;
+    int offset = IMM;
+    offset <<=2;
+    uint64_t rs = RS;
+    uint64_t rt = RT;
 
-    //EmotionEngine::REG(id) gets the name of the register associated with an id.
-    output << EmotionEngine::REG(reg1) << ", ";
-    if (!reg2)
+    output << EmotionEngine::REG(rs) << ", ";
+    if (!rt)
+    {
```

Comment:
Only use curly braces for control-flow blocks that span more than one line.

e.g.
```
if (cond)
    do_something();

if (cond2)
{
    do_something();
    do_something_else();
}